client/src/controllers/TabsController.test.js

```diff
@@ -0,0 +1,256 @@
+import { Application } from '@hotwired/stimulus';
+import { TabsController } from './TabsController';
+
+function simulateKeydown(key) {
+  const event = new KeyboardEvent('keydown', {
+    key,
+    code: key,
+    bubbles: true,
+    cancelable: true,
+  });
+  document.activeElement.dispatchEvent(event);
+}
+
+jest.useFakeTimers();
+
+describe('TabsController', () => {
+  let app;
+  const oldWindowLocation = window.location;
+
+  const setup = async (html) => {
+    document.body.innerHTML = `<main>${html}</main>`;
+
+    app = Application.start();
+    app.register('w-tabs', TabsController);
+    await Promise.resolve();
+  };
```

Comment:
Instead of setting the HTML in each grouping of tests, start with a reasonable full HTML that can be used for 'most' tests and then you do not have to write it each time.

Here's a modified starting HTML with the following changes

1. Let's have more fun, I have used coffee/chocolate/cheese but feel free to adjust.
2. Let's have three or maybe four tabs, this makes it easier to test different cases.
3. The keyboard actions do not need to be on each label (I think), as the keyboard events bubble up, this is also going to be a simpler set of HTML in the DOM.
4. Fixed up spacing/nesting, using only 2 spaces not a mix of 2 & 4
5. Set up a second value `identifier` to the default `w-tabs`, just makes it easier to change this in tests if needed.


```suggestion
  const setup = async (
    html = `
  <div class="w-tabs" data-controller="w-tabs" data-action="popstate@window->w-tabs#loadHistory" data-w-tabs-selected-class="animate-in">
    <div class="w-tabs__list" role="tablist" data-w-tabs-target="list" data-action="click->w-tabs#handleTabChange:prevent keydown.right->w-tabs#selectNext keydown.left->w-tabs#selectPrevious keydown.home->w-tabs#selectFirst keydown.end->w-tabs#selectLast">
      <a id="tab-label-tab-1" href="#tab-tab-1"role="tab" tabindex="-1" data-w-tabs-target="label">
        Cheese
      </a>
      <a id="tab-label-tab-2" href="#tab-tab-2" role="tab">
        Chocolate
      </a>
      <a id="tab-label-tab-3" href="#tab-tab-3" role="tab">
        Coffee
      </a>
    </div>
    <div class="tab-content tab-content--comments-enabled">
      <section id="tab-tab-1" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-1" data-w-tabs-target="panel">
        All about cheese
      </section>
      <section id="tab-tab-2" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-2" data-w-tabs-target="panel">
        All about chocolate
      </section>
      <section id="tab-tab-3" class="w-tabs__panel " role="tabpanel" aria-labelledby="tab-label-tab-3" data-w-tabs-target="panel">
        All about coffee
      </section>
    </div>
  </div>
  `,
    identifier = 'w-tabs',
  ) => {
    document.body.innerHTML = `<main>${html}</main>`;

    app = Application.start();
    app.register(identifier, TabsController);
    await Promise.resolve();
  };
```